### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by AndrewDryga**

### CODE REVIEW

File: lib/mithril_api/client_api/client.ex [hunk 1]
```diff
@@ -11,8 +11,7 @@ defmodule Mithril.ClientAPI.Client do
     field :settings, :map, default: %{}
 
     belongs_to :client_type, Mithril.ClientTypeAPI.ClientType
-    # TODO: Remove Web prefix
-    belongs_to :user, Mithril.Web.UserAPI.User
+    belongs_to :user, Mithril.UserAPI.User
```
Comment (AndrewDryga) [inline lib/mithril_api/client_api/client.ex]: TODO found

File: lib/mithril_api/authorization/token.ex [hunk 2]
```diff
@@ -8,8 +8,7 @@ defmodule Mithril.Authorization.Token do
   alias Mithril.Authorization.GrantType.AuthorizationCode
   alias Mithril.Authorization.GrantType.RefreshToken
 
-  # TODO:
-  # rename grant_type to response_type
+  # TODO: rename grant_type to response_type
```
Comment (AndrewDryga) [inline lib/mithril_api/authorization/token.ex]: TODO found

File: lib/mithril_api/authorization/grant_type/authorization_code.ex [hunk 3]
```diff
@@ -75,27 +74,12 @@ defmodule Mithril.Authorization.GrantType.AuthorizationCode do
...
-    if Mithril.Utils.List.subset?(scopes, required_scopes) do
-      {:ok, token}
-    else
-      GrantTypeError.invalid_scope(scopes)
-    end
-  end
-
   defp validate_token_is_not_used({:error, err, code}), do: {:error, err, code}
-  defp validate_token_is_not_used({:ok, token}) do
+  defp validate_token_is_not_used({:ok, token, app}) do
```
Comment (AndrewDryga) [inline lib/mithril_api/authorization/grant_type/authorization_code.ex]: TODO found

Comment (AndrewDryga) [inline lib/mithril_api/authorization/grant_type/authorization_code.ex]: Found a TODO tag in a comment: # TODO: Probably no need to do this at all. When client exchanges code for token,

File: lib/mithril_api/authorization/token.ex (hunk 2 above)
Comment (AndrewDryga) [inline lib/mithril_api/authorization/token.ex]: Found a TODO tag in a comment: # TODO:

File: lib/mithril_api/client_api/client.ex (hunk 1 above)
Comment (AndrewDryga) [inline lib/mithril_api/client_api/client.ex]: Found a TODO tag in a comment: # TODO: Remove Web prefix

File: lib/mithril_api/authorization/grant_type/refresh_token.ex [hunk 4]
```diff
@@ -66,13 +66,13 @@ defmodule Mithril.Authorization.GrantType.RefreshToken do
...
   defp create_access_token({:error, _, _} = error), do: error
-  defp create_access_token({:ok, client, token, app}) do
+  defp create_access_token({:ok, client, token, _app}) do
     Mithril.TokenAPI.create_access_token(%{
       user_id: token.user_id,
       details: %{
         grant_type: "refresh_token",
         client_id: client.id,
-        scope: app.scope
+        scope: token.details["scope"]
```
Comment (gmile) [inline lib/mithril_api/authorization/grant_type/refresh_token.ex]: Why taking `scope` from token and not from app?

File: test/unit/authorization/grant_type_auth_code_test.exs [hunk 5]
```diff
@@ -31,7 +30,7 @@ defmodule Mithril.Authorization.GrantType.AuthorizationCodeTest do
     assert token.details.refresh_token
     assert token.details.grant_type == "authorization_code"
     assert token.details.redirect_uri == client.redirect_uri
-    assert token.details.scope == "legal_entity:read"
+    assert token.details.scope == "legal_entity:read legal_entity:write"
```
Comment (gmile) [inline test/unit/authorization/grant_type_auth_code_test.exs]: Where did `legal_entity:write` come from?

File: test/web/controllers/auth/apps_controller_test.exs [hunk 6]
```diff
@@ -91,7 +91,7 @@ defmodule Mithril.OAuth.AppControllerTest do
 
     assert app.user_id == user.id
     assert app.client_id == client.id
-    assert app.scope == "legal_entity:read legal_entity:write"
```
Comment (gmile) [inline test/web/controllers/auth/apps_controller_test.exs]: Where did `legal_entity:read` go?

File: config/config.exs [hunk 7]
```diff
@@ -68,35 +68,6 @@ config :logger_json, :backend,
 
 import_config "#{Mix.env}.exs"
 
-config :mithril_api, :scopes, ~w(
```
Comment (gmile) [inline config/config.exs]: Were these already extracted somewhere? Can you point to a place in code / k8s config?

**Comment by gmile** [comment]
Overall looks like a leap in the right direction. Just want to clarify a couple of questions (e.g. nothing has been missed) before merging this

File: config/config.exs (hunk 7 above)
Comment (AndrewDryga) [inline config/config.exs]: Nope, and there are no Confex usage in here so it is not possible to override this values in ENV.

File: lib/mithril_api/authorization/grant_type/refresh_token.ex (hunk 4 above)
Comment (AndrewDryga) [inline lib/mithril_api/authorization/grant_type/refresh_token.ex]: Refreshed token should have exactly the same scope as it had before, if oAuth client wants to extend that - he needs to request the new token.

File: config/config.exs (hunk 7 above)
Comment (gmile) [inline config/config.exs]: @AndrewDryga so what will happen if we deploy this? If this is not going to break, the where is the list of scopes to look at right now?

File: test/unit/authorization/grant_type_auth_code_test.exs (hunk 5 above)
Comment (AndrewDryga) [inline test/unit/authorization/grant_type_auth_code_test.exs]: Token scope equals to approval scope, see line 13. Now you can't ask for specific scope on code_grant auth.

File: test/web/controllers/auth/apps_controller_test.exs (hunk 6 above)
Comment (AndrewDryga) [inline test/web/controllers/auth/apps_controller_test.exs]: Approval did not authorize to make writes, this was a security issue.

**Comment by gmile** [comment]
One note: it is not ABSOLUTELY required for mithril in all environments to have proper `client_types` and `roles`. Someone have to double-check that all envs have roles / clien_types according to this doc: https://edenlab.atlassian.net/wiki/display/EH/Scopes+model

**Pull request merged**